api/admin/controller/library_settings.py

```diff
@@ -216,29 +221,83 @@ def get_extra_geographic_information(self, value):
+    def library_configuration_settings(
+            self, library, validators_by_format, settings=None
+    ):
+        """Validate and update a library's configuration settings based on incoming new
+        values.
+
+        :param library: A Library
+        :param validators_by_format: A dictionary mapping the 'format' field from a setting
+           configuration to a corresponding validator object.
+        :param settings: A list of setting configurations to use in tests instead of
+           Configuration.LIBRARY_SETTINGS
+        """
+        settings = settings or Configuration.LIBRARY_SETTINGS
+        for setting in settings:
+            # Validate the incoming value.
+            validator = None
+            if 'format' in setting:
+                validator = validators_by_format.get(setting["format"])
+            validated_value = self._validate_setting(library, setting, validator)
+
+            if isinstance(validated_value, ProblemDetail):
+                # Validation failed -- return a ProblemDetail.
+                return validated_value
+
+            # Validation succeeded -- set the new value.
+            ConfigurationSetting.for_library(setting['key'], library).value = validated_value
+
+    def _validate_setting(self, library, setting, validator=None):
+        """Validate the incoming value for a single library setting.
+
+        :param library: A Library
+        :param setting: Configuration data for one of the library's settings.
+        :param validator: A validation object for data of this type.
+        """
+        # TODO: there are some opportunities for improvement here:
+        # * There's no standard interface for validators.
+        # * We can handle settings that are lists of certain types (language codes,
+        #   geographic areas), but not settings that are a single value of that type
+        #   (_one_ language code or geographic area). Sometimes there's even an implication
+        #   that a certain data type ('geographic') _must_ mean a list.
+        # * A list value is returned as a JSON-encoded string. It
+        #   would be better to keep that as a list for longer in case
+        #   controller code needs to look at it.
+        format = setting.get('format')
+        type = setting.get('type')
+
+        # In some cases, if there is no incoming value we can use a
+        # default value or the current value.
+        #
+        # When the configuration item is a list, we can't do this
+        # because an empty list may be a valid value.
+        current_value = self.current_value(setting, library)
+        default_value = setting.get('default') or current_value
+
+        if format == "geographic":
+            value = self.list_setting(setting)
+            value = validator.validate_geographic_areas(value, self._db)
+        elif format == "announcements":
+            value = self.list_setting(setting)
```

Comment:
The announcements aren't in quite the same format as the geographic input; I'm not sure we'll want/need to run them through `list_setting`.  I'm envisioning them being a list of dictionaries, each with "content", "start_date", and "end_date" keys.  Whereas the geographic input--if I'm remembering correctly!--is just a list of strings.

Reply:
I see. In that case I think list_setting is going to get us a list of JSON-encoded strings, which we'll need to decode. Probably best to pass in a flag and do this inside `list_setting`.